### PR TITLE
Validate Workload admission

### PR DESCRIPTION
**Pull request by kerthcet**
Signed-off-by: kerthcet <kerthcet@gmail.com>

#### What type of PR is this?
/kind feature

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes-sigs/kueue/issues/328

#### Special notes for your reviewer:
Ignore the validation of `Each resourceName should be listed in the pod spec, although it might be too cumbersome to validate, so we could ignore it.`

### CODE REVIEW

File: apis/kueue/webhooks/workload_webhook.go [hunk 1]
```diff
@@ -161,6 +166,30 @@ func validatePodSetName(name string, fldPath *field.Path) field.ErrorList {
...
+	var allErrs field.ErrorList
+	allErrs = append(allErrs, validateNameReference(string(admission.ClusterQueue), path.Child("clusterQueue"))...)
+
+	flavorPath := path.Child("podSetFlavors")
+	for i, flavor := range admission.PodSetFlavors {
+		allErrs = append(allErrs, validateNameReference(flavor.Name, flavorPath.Index(i).Child("name"))...)
+	}
+
+	names := sets.NewString()
+	for _, flavor := range obj.Spec.PodSets {
```
Comment (alculquicondor) [inline apis/kueue/webhooks/workload_webhook.go:180]:
```suggestion
	for _, ps := range obj.Spec.PodSets {
```

File: apis/kueue/webhooks/workload_webhook.go [hunk 2]
```diff
@@ -161,6 +166,30 @@ func validatePodSetName(name string, fldPath *field.Path) field.ErrorList {
...
+		names.Insert(flavor.Name)
+	}
+
+	for i, flavor := range obj.Spec.Admission.PodSetFlavors {
```
Comment (alculquicondor) [inline apis/kueue/webhooks/workload_webhook.go:184]:
```suggestion
	for i, ps := range obj.Spec.Admission.PodSetFlavors {
```

File: pkg/util/testing/wrappers.go [hunk 3]
```diff
@@ -222,16 +222,30 @@ func (w *WorkloadWrapper) NodeSelector(kv map[string]string) *WorkloadWrapper {
...
+				Name:    kueue.DefaultPodSetName,
 				Flavors: make(map[corev1.ResourceName]string),
 			},
-		},
-	}}
+		}
+		return wrap
+	}
+
+	var flavors []kueue.PodSetFlavors
```
Comment (alculquicondor) [inline pkg/util/testing/wrappers.go:240]:
```suggestion
	var psFlavors []kueue.PodSetFlavors
```

File: apis/kueue/webhooks/workload_webhook.go [hunk 4]
```diff
@@ -161,6 +166,30 @@ func validatePodSetName(name string, fldPath *field.Path) field.ErrorList {
...
 }
 
+func validateAdmission(obj *kueue.Workload, path *field.Path) field.ErrorList {
+	admission := obj.Spec.Admission
+	var allErrs field.ErrorList
+	allErrs = append(allErrs, validateNameReference(string(admission.ClusterQueue), path.Child("clusterQueue"))...)
+
+	flavorPath := path.Child("podSetFlavors")
+	for i, flavor := range admission.PodSetFlavors {
+		allErrs = append(allErrs, validateNameReference(flavor.Name, flavorPath.Index(i).Child("name"))...)
```
Comment (alculquicondor) [inline apis/kueue/webhooks/workload_webhook.go:176]:
```suggestion
		allErrs = append(allErrs, validatePodSetName(flavor.Name, flavorPath.Index(i).Child("name"))...)
```

Although, maybe we don't need this validation, given than the value should be present in the podSets, which are already validated.

Comment (kerthcet) [reply apis/kueue/webhooks/workload_webhook.go:176]: Yes, updated.

**Comment by alculquicondor** [comment]
/approve

Can you squash?

**Comment by kerthcet** [comment]
Squashed

**Comment by alculquicondor** [comment]
Great!
/lgtm

**Pull request merged**